src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
+		if (!a20.enable_global) {
+			result = a20_local_enable();
+			if (result == Result::OK) {
+				a20.enable_global = true;
+			}
+		}
+		set_return_value(result);
 		break;
-		
-	case XMS_GLOBAL_ENABLE_A20:									/* 03 */
-	case XMS_LOCAL_ENABLE_A20:									/* 05 */
-		SET_RESULT(XMS_EnableA20(true));
+	case 0x04: // Global Disable A20
+		// This appears to be how Microsoft HIMEM.SYS implements this
+		if (a20.enable_global) {
+			result = a20_local_disable();
+			if (result == Result::OK) {
+				a20.enable_global = false;
+			}
+		}
+		set_return_value(result);
 		break;
-	case XMS_GLOBAL_DISABLE_A20:								/* 04 */
-	case XMS_LOCAL_DISABLE_A20:									/* 06 */
-		SET_RESULT(XMS_EnableA20(false));
-		break;	
-	case XMS_QUERY_A20:											/* 07 */
-		reg_ax = XMS_GetEnabledA20();
+	case 0x05: // Local Enable A20
+		set_return_value(a20_local_enable());
+		break;
+	case 0x06: // Local Disable A20
+		set_return_value(a20_local_disable());
+		break;
+	case 0x07: // Query A20
+		reg_ax = a20_is_enabled() ? 1 : 0;
 		reg_bl = 0;
 		break;
-	case XMS_QUERY_FREE_EXTENDED_MEMORY:						/* 08 */
-		reg_bl = XMS_QueryFreeMemory(reg_ax,reg_dx);
+	case 0x08: // Query Free Extended Memory
+		result = xms_query_free_memory(reg_eax, reg_edx);
+		reg_bl = static_cast<uint8_t>(result);
+		// Cap sizes for older programs; newer ones use function 0x88
+		reg_eax = std::min(reg_eax, static_cast<uint32_t>(UINT16_MAX));
+		reg_edx = std::min(reg_edx, static_cast<uint32_t>(UINT16_MAX));
 		break;
-	case XMS_ALLOCATE_ANY_MEMORY:								/* 89 */
-		reg_edx &= 0xffff;
-		[[fallthrough]];
-	case XMS_ALLOCATE_EXTENDED_MEMORY:							/* 09 */
-		{
+	case 0x09: // Allocate Extended Memory Block
+	{
 		uint16_t handle = 0;
-		SET_RESULT(XMS_AllocateMemory(reg_dx,handle));
+		set_return_value(xms_allocate_memory(reg_dx, handle));
 		reg_dx = handle;
-		}; break;
-	case XMS_FREE_EXTENDED_MEMORY:								/* 0a */
-		SET_RESULT(XMS_FreeMemory(reg_dx));
+	} break;
+	case 0x0a: // Free Extended Memory Block
+		set_return_value(xms_free_memory(reg_dx));
 		break;
-	case XMS_MOVE_EXTENDED_MEMORY_BLOCK:						/* 0b */
-		SET_RESULT(XMS_MoveMemory(SegPhys(ds)+reg_si),false);
+	case 0x0b: // Move Extended Memory Block
+		result = xms_move_memory(SegPhys(ds) + reg_si);
+		set_return_value_bl_only_fail(result);
 		break;
-	case XMS_LOCK_EXTENDED_MEMORY_BLOCK: {						/* 0c */
-		uint32_t address;
-		Bitu res = XMS_LockMemory(reg_dx, address);
-		if(res) reg_bl = (uint8_t)res;
-		reg_ax = (res==0);
-		if (res==0) { // success
-			reg_bx=(uint16_t)(address & 0xFFFF);
-			reg_dx=(uint16_t)(address >> 16);
-		};
-		}; break;
-	case XMS_UNLOCK_EXTENDED_MEMORY_BLOCK:						/* 0d */
-		SET_RESULT(XMS_UnlockMemory(reg_dx));
+	case 0x0c: // Lock Extended Memory Block
+	{
+		uint32_t address = 0;
+		result           = xms_lock_memory(reg_dx, address);
+		set_return_value(result);
+		if (result == Result::OK) {
+			// success
+			reg_bx = RealOff(address);
+			reg_dx = RealSeg(address);
```

Comment:
I wish these were called RealOffset and RealSegment...

Reply:
Done. Note: this was mostly a search&replace commit, with some small exceptions no other cleanups were done.